server/komodo-rest/src/main/java/org/komodo/servicecatalog/TeiidOpenShiftClient.java

```diff
@@ -632,213 +666,258 @@ private DeploymentCondition getDeploymentConfigStatus(DeploymentConfig dc) {
-                        work.status = Status.FAILED;
-                        work.statusMessage = build.getStatus().getMessage();
-                        KLog.getLogger().debug("Build failed :" + work.buildName + ". Reason "
-                                + build.getStatus().getLogSnippet());
-                    }
 
-                    synchronized (work) {
-                        work.lastUpdated = System.currentTimeMillis();
-                        workQueue.poll(); // remove
-                        if (shouldReQueue) {
-                            workQueue.offer(work); // add at end
+                        synchronized (work) {
+                            work.setLastUpdated();
+                            workQueue.poll(); // remove
+                            if (shouldReQueue) {
+                                workQueue.offer(work); // add at end
+                            } else {
+                                //
+                                // dispose of the publish config artifacts
+                                //
+                                work.publishConfiguration().dispose();
+                            }
                         }
                     }
+                } finally {
+                    running = false;
+                    kubernetesClient.close();
                 }
-            } finally {
-                running = false;
-                kubernetesClient.close();
             }
-        }});
+        });
     }
 
-    /**
-     * Publish the vdb as a virtualization
-     *
-     * @param authToken the authentication token for Openshift
-     * @param uow the transaction for accessing the vdb's properties
-     * @param vdb the vdb for virtualising
-     * @return the build status of the virtualization
-     * @throws KException if error occurs
-     */
-    public BuildStatus publishVirtualization(AuthToken authToken, UnitOfWork uow, PublishConfiguration publishConfig) throws KException {
-        String namespace = ApplicationProperties.getNamespace();
-        Config config = new ConfigBuilder().build();
-        OpenShiftConfig.wrap(config).setBuildTimeout(publishConfig.buildTimeoutInSeconds);
-        KubernetesClient kubernetesClient = new DefaultKubernetesClient(config);
-        final OpenShiftClient client = kubernetesClient.adapt(OpenShiftClient.class);
+    protected void configureBuild(BuildStatus work) {
+        //
+        // Sets the status immediately since the task may have to wait
+        // if all threads are currently undergoing tasks. This ensures
+        // that the monitor thread will not try to do anything more with it.
+        //
+        work.setStatus(Status.CONFIGURING);
 
-        Vdb vdb = publishConfig.vdb;
-        String vdbName = vdb.getVdbName(uow);
-        try {
-            debug("Publishing (" + vdbName + ") - Start publishing of virtualization: " + vdbName);
+        threadService.execute(new Runnable() {
```

Comment:
Why another thread for the build? Was there a reason?

Reply:
Configure takes a while. So as it is now initiated by the monitor thread, it would hold up the monitor thread from monitoring all the other builds. With 4 threads in the pool, up to 3 configures can be done in parallel along with the monitor thread monitoring. Thus, avoiding potential bottleneck.

Reply:
At a given request, there will be at most one configure thread + monitoring thread + calling thread in play. Previously I had the calling thread do the configure, now there will be one extra, which IMO not really needed because the aim is to free the calling thread and do everything in the secondary thread thus being ASYNC. Monitoring thread is also ephemeral once the build task is done it is terminated.

If a different publish request comes in the process that will get its own executor service and it's own Threads, as this is JEE servlet process is request/response model. The flaw as I can see even from the original design is there will be verbose monitoring threads if there are concurrent publish jobs running, we could tighten to only look at the work they are initiated with, not every work in the queue.

The status request should look at the shared status object, and return the result in the same calling thread, which it does.

Reply:
Calling thread has 30 second timeout being an http request so has to start things off then return.

Monitoring thread is ephemeral but there will always be 1 as long as there is work on the workQueue.

So there is not 1 extra since there remain 2 threads to do the work but both are not constrained by any timeouts and neither is constrained by the other in that the monitoring thread is free to continue monitoring other tasks as well as the configuring the other thread is performing.

None of the threads yield to each other so if there were multiple executor services, sooner or later the threads would get bottled-up waiting for a cpu slot. So better to limit the executor services to 1 and the number of threads to a sensible number. The executor service takes care of queueing tasks internally if a thread is not currently available.

The parallel instantiation of multiple monitor threads is an issue and better then to take this out of the executor service and keep as a single running background thread (sleeps if there is nothing in the workqueue).